scribus/plugins/import/pdf/slaoutput.cpp

```diff
@@ -3896,3 +3926,680 @@ bool SlaOutputDev::checkClip()
+							,0,100
+							,0,100
+							,0,0
+							,0,0
+	};
+	boundingBoxShape.svgMoveTo(bbosdoubles[0], bbosdoubles[1]);
+	for (int a = 0; a < 16; a += 2)
+	{
+		boundingBoxShape.append(FPoint(bbosdoubles[a * 2], bbosdoubles[a * 2 + 1]));
+	}
+	boundingBoxShape.scale(textNode->width() / 100.0, textNode->height() / 100.0);
+
+	textNode->SetFrameShape(32, bbosdoubles);
+	textNode->ContourLine = textNode->PoLine.copy();
+
+	m_doc->Items->removeLast();
+	m_Elements->append(textNode);
+	if (m_groupStack.count() != 0)
+	{
+		m_groupStack.top().Items.append(textNode);
+		applyMask(textNode);
+	}
+}
+
+/*code mostly taken from importodg.cpp which also supports some line styles and more fill options etc...*/
+//FIXME: This needs to be implemented based on the framework
+void SlaOutputDev::finishItem(PageItem* item)
+{
+	item->ClipEdited = true;
+	item->FrameType = 3;
+	//this requires that PoLine is set
+	//FPoint wh = getMaxClipF(&item->PoLine);
+	//item->setWidthHeight(wh.x(), wh.y());
+	//item->Clip = flattenPath(item->PoLine, item->Segments);
+	item->OldB2 = item->width();
+	item->OldH2 = item->height();
+	item->updateClip();
+	item->OwnPage = m_doc->OnPage(item);
+	//item->setFillTransparency(1.0 - state->getFillOpacity() > state->getStrokeOpacity() ? state->getFillOpacity() : state->getStrokeOpacity());
+	//item->setLineTransparency(1.0);
+}
+
+TextFramework::TextFramework()
+{
+	m_textRegions.push_back(activeTextRegion);
+	setCharMode(AddCharMode::ADDFIRSTCHAR);
+}
+
+TextFramework::~TextFramework()
+{
+}
+
+void TextFramework::addNewTextRegion()
+{
+	activeTextRegion = TextRegion();
+	m_textRegions.push_back(activeTextRegion);
+	setCharMode(TextFramework::AddCharMode::ADDFIRSTCHAR);
+}
+
+void TextFramework::addChar(GfxState* state, double x, double y, double dx, double dy, double originX, double originY, CharCode code, int nBytes, Unicode const* u, int uLen)
+{
+	(this->*(m_addChar))(state, x, y, dx, dy, originX, originY, code, nBytes, u, uLen);
+}
+
+bool TextFramework::isNewLineOrRegion(QPointF newPosition)
+{
+	return (activeTextRegion.coLinera(activeTextRegion.lastXY.y(), activeTextRegion.textRegionLines.back().baseOrigin.y()) &&
+		!activeTextRegion.coLinera(newPosition.y(), activeTextRegion.lastXY.y()))
+		|| (activeTextRegion.coLinera(newPosition.y(), activeTextRegion.lastXY.y())
+			&& !activeTextRegion.closeToX(newPosition.x(), activeTextRegion.lastXY.x()));
+}
+
+PdfGlyph TextFramework::AddFirstChar(GfxState* state, double x, double y, double dx, double dy, double originX, double originY, CharCode code, int nBytes, Unicode const* u, int uLen)
+{
+	//qDebug() << "AddFirstChar() '" << u << " : " << uLen;	
+	PdfGlyph newGlyph = TextFramework::AddCharCommon(state, x, y, dx, dy, u, uLen);
+	activeTextRegion.glyphs.push_back(newGlyph);
+	setCharMode(AddCharMode::ADDBASICCHAR);
+
+	//only need to be called for the very first point
+	if (activeTextRegion.addGlyphAtPoint(QPointF(x, y), newGlyph) == TextRegion::FrameworkLineTests::FAIL)

```

Comment:
somehow, i don't like this... side effects in an if conditions seem to be dangerous to me.

i would split it in two lines and if `addGlyphAtPoint` return value seems to be if it has failed or not, i would change its signature to return true or false (and check inside of it if the state is a fail or not)

```
auto success = add();
if (!success)
  qDebug("...");

Reply:
ok fair point